### PR TITLE
feat: Avm/res/network/load balancer

### LINKED ISSUE #191: Bicep Landing Zone (Subscription) Vending module
### Have you checked this module does not already exist in the registry?

Yes

### Why is the module needed?

Customers today have a gap in being able to create subscriptions programmatically with the base networking resources they generally require in a repeatable fashion. 

We in the CAE team have created a Bicep Landing Zone vending module (https://github.com/Azure/bicep-lz-vending/), as well as a Terraform equivalent (HTTPS://aka.ms/LZ-vending/TF), to address this gap. 

We would like to have it consumable in the same way in bicep as it is via terraform in a public module registry to remove an additional step for consumers to have to worry about.  

### Module path

lz/subWithNetworking

### Describe the module

All details can be found in repo here https://github.com/Azure/bicep-lz-vending

### PR DESCRIPTION
**Pull request by arnoldna**
## Description

Initial Bicep Load Balancer module Closes #191 

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

Enable GitHub Worksflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) as a secret in your forked repository called `PAT`.

## Adding a new module

- [X ] A proposal has been submitted and approved.
- [ X] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [ X] I have run `brm validate` locally to verify the module files.
- [X ] I have run deployment tests locally to ensure the module is deployable.

## Updating an existing module

- [ ] This is a bug fix:
  - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
  - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
- [ ] I have run `brm validate` locally to verify the module files.
- [X ] I have run deployment tests locally to ensure the module is deployable.
- [ X] I have read the [Updating an existing module](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md#updating-an-existing-module) section in the contributing guide and updated the `version.json` file properly:
  - [ ] The PR contains backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`.
  - [ ] The PR contains backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] The PR contains breaking changes, and I have bumped the MAJOR version in `version.json`.
- [ X] I have updated the examples in README with the latest module version number.

### CODE REVIEW

File: avm/res/network/load-balancer/backend-address-pool/README.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+- Required: Yes
+- Type: string
+
+### Parameter: `syncMode`
+
+Backend address synchronous mode for the backend pool.
+- Required: No
+- Type: string
+- Default: `''`
+- Allowed: `['', Automatic, Manual]`
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/backend-address-pool/README.md:81]: It apperas that the used readme utility was not the latest as this line shoud now be split across multiple lines

File: avm/res/network/load-balancer/backend-address-pool/main.bicep [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  'Manual'
+])
+@description('Optional. Backend address synchronous mode for the backend pool.')
+param syncMode string = ''
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
+
+resource avmTelemetry 'Microsoft.Resources/deployments@2023-07-01' = if (enableTelemetry) {
+  name: '46d3xbcp.res.network.loadbalancer.${replace('-..--..-', '.', '-')}.${substring(uniqueString(deployment().name), 0, 4)}'
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/backend-address-pool/main.bicep:32]:
```suggestion
  name: '46d3xbcp.res.network-loadbalancer.${replace('-..--..-', '.', '-')}.${substring(uniqueString(deployment().name), 0, 4)}'
```

File: avm/res/network/load-balancer/backend-address-pool/version.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "$schema": "https://aka.ms/bicep-registry-module-version-file-schema#",
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/backend-address-pool/version.json:2]: Please remove the `version.json` files from all child modules

File: avm/res/network/load-balancer/backend-address-pool/main.bicep [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  'Automatic'
+  'Manual'
+])
+@description('Optional. Backend address synchronous mode for the backend pool.')
+param syncMode string = ''
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
+
+resource avmTelemetry 'Microsoft.Resources/deployments@2023-07-01' = if (enableTelemetry) {
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/backend-address-pool/main.bicep:31]: Please remove any reference of telemetry from child modules

Comment (AlexanderSehr) [reply avm/res/network/load-balancer/backend-address-pool/main.bicep:31]: Including parameters etc.

File: avm/res/network/load-balancer/main.bicep [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+  'Role Based Access Control Administrator (Preview)': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'f58310d9-a9f6-439a-9e8d-f62e7b41a168')
+  'User Access Administrator': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '18d7d88d-d35e-4fb5-a5c3-7773c20a72d9')
+}
+
+// ============ //
+// Dependencies //
+// ============ //
+
+resource avmTelemetry 'Microsoft.Resources/deployments@2023-07-01' = if (enableTelemetry) {
+  name: '46d3xbcp.res.network.loadbalancer.${replace('-..--..-', '.', '-')}.${substring(uniqueString(deployment().name, location), 0, 4)}'
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/main.bicep:153]:
```suggestion
  name: '46d3xbcp.res.network-loadbalancer.${replace('-..--..-', '.', '-')}.${substring(uniqueString(deployment().name, location), 0, 4)}'
```

File: avm/res/network/load-balancer/main.bicep [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+
+type roleAssignmentType = {
+  @description('Required. The name of the role to assign. If it cannot be found you can specify the role definition ID instead.')
+  roleDefinitionIdOrName: string
+
+  @description('Required. The principal ID of the principal (user/group/identity) to assign the role to.')
+  principalId: string
+
+  @description('Optional. The principal type of the assigned principal ID.')
+  principalType: ('ServicePrincipal' | 'Group' | 'User' | 'ForeignGroup' | 'Device' | null)?
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/main.bicep:331]:
```suggestion
  principalType: ('ServicePrincipal' | 'Group' | 'User' | 'ForeignGroup' | 'Device')?
```

File: test.txt
Comment (AlexanderSehr) [inline test.txt:1]: The hell? :D

File: avm/res/network/load-balancer/README.md [hunk 7]
```diff
@@ -0,0 +1,1193 @@
...
+
+This module deploys a Load Balancer.
+
+## Navigation
+
+- [Resource Types](#Resource-Types)
+- [Usage examples](#Usage-examples)
+- [Parameters](#Parameters)
+- [Outputs](#Outputs)
+- [Cross-referenced modules](#Cross-referenced-modules)
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/README.md:11]: The [notes](https://github.com/Azure/ResourceModules/tree/main/modules/network/load-balancer) of the original Readme got lost. Was that intentional?

If not, I'd recommend to copy the readme from CARML and re-run the readme generation (which should preserve the notes).

File: avm/res/network/load-balancer/tests/e2e/defaults/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'dep-${namePrefix}-network.loadbalancers-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'nlbdef'
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/tests/e2e/defaults/main.test.bicep:15]:
```suggestion
param serviceShort string = 'nlbmin'
```

Comment (AlexanderSehr) [reply avm/res/network/load-balancer/tests/e2e/defaults/main.test.bicep:15]: Requires a regeneration of the top-level readme

File: test.txt
Comment (arnoldna) [reply test.txt:1]: I thought I deleted that file. It is gone now

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**